---
title: "dnmt3a_tet2_mutated_week30"
author: "Sanketh"
date: "2024-01-04"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Setting the directory and load required packages  

```{r}
rm(list=ls())
library(tidyverse)
library(Seurat)
library(ggplot2)
library(stringr)
#install.packages("Matrix")
library(Matrix)
library(scCustomize)
library(harmony)
library(devtools)
library(presto)
#devtools::install_github('immunogenomics/presto')
setwd("/Users/sanketh/Desktop/proj1_tet_dnmt3a_mut/")
getwd()
#Set the data dir to week 30 
dataDir <- ("/Users/sanketh/Desktop/proj1_tet_dnmt3a_mut/data/week30/")
```


```{r}
read_counts <- function(file){
  x <- readMM(file) %>%
  as.matrix()
  sample <- basename(file)
  sample <- str_replace(sample, "_matrix.*", "")
  genes_x <- read.table(file = paste0(dataDir,sample,"_genes.tsv.gz"), header = F, sep = "\t")
  genes <- genes_x[,2]
  gene_names <- make.unique(genes) #This is a built in function in r to make the gene names unique 
  barcodes_x <- read.table(file = paste0(dataDir,sample,"_barcodes.tsv.gz"), header = F, sep = "\t")
  barcodes <- paste0(barcodes_x[,1],"_",sample)
  row.names(x) <- gene_names
  colnames(x) <- barcodes 
  return(x)
}
```

```{r}
read_barcodes <- function(file){
  y <- read.table(file, header = F, sep = "\t") %>%
    as.data.frame() %>%
    rename(Cellbarcode = "V1")
  sample <- basename(file)
  sample <- str_replace(sample,"_barcodes.tsv.gz","")
  barcode_metadata <- str_split(sample,pattern = "_") %>%
    unlist() %>%
    purrr::set_names(c("Sample_geo","week","mutation","type","subtype")) %>%
    as.list()
  y$sample_geo <- rep(barcode_metadata$Sample_geo,nrow(y))
  y$week <- rep(barcode_metadata$week,nrow(y))
  y$mutation <- rep(barcode_metadata$mutation,nrow(y))
  y$genotype <- rep(paste(barcode_metadata$mutation,barcode_metadata$type,
                           barcode_metadata$subtype, sep="_"),nrow(y))
  return(y)
}
```



# Basic function to convert human to mouse gene names

```{r}
convert_human_to_mouse <- function(gene_list){
  
  output = c()
  
  for(gene in gene_list){
    class_key = (mouse_human_genes %>% filter(Symbol == gene & Common.Organism.Name=="human"))[['DB.Class.Key']]
    if(!identical(class_key, integer(0)) ){
      mouse_genes = (mouse_human_genes %>% filter(DB.Class.Key == class_key & Common.Organism.Name=="mouse, laboratory"))[,"Symbol"]
      for(mouse_gene in mouse_genes){
        output = append(output,mouse_gene)
      }
    }
  }
  
  return (output)
}
```
  
  
```{r}
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
mouse_human_genes = read.csv("http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt",sep="\t")
m.s.genes <- convert_human_to_mouse(cc.genes.updated.2019$s.genes)
m.g2m.genes <- convert_human_to_mouse(cc.genes.updated.2019$g2m.genes)
```



```{r}
#Reading in the count matrices
counts_files <- list.files(path = "./data/week30/", full.names = T, pattern = "*.mtx.gz")
samples <- map_chr(counts_files,basename)
samples <- str_replace(samples, "_matrix.*", "")
names(counts_files) <- samples
counts <- purrr::map(counts_files,read_counts) #Running now on all samples 
```

```{r}
##Write code to create metadata for the cells : 
barcode_files <- list.files (path = "./data/week30/", full.names = T, pattern = "*barcodes.tsv.gz")
names(barcode_files) <- samples
metadata_cells <- purrr::map(barcode_files,read_barcodes)
```

#Creating Seurat Object 
```{r}
library(Matrix)
objs <- purrr::map2(counts, metadata_cells,
                    ~CreateSeuratObject(counts = .x,
                                        meta.data = .y,
                                        )) #Can ignore the warning as Surat directly converts into sparse matrix which is req. 
```

```{r}
merged_seurat <- purrr::reduce(objs, function(x,y) {merge(x,y)})
#Once the merge is done the counts are not merged hence this needs to be 
#to merge the counts from different samples into one counts layer 
merged_seurat <- JoinLayers(merged_seurat)
merged_seurat
```

# Filter, normalize, scale, linear dimensionality reduction
```{r}
# Filter, normalize, scale, linear dimensionality reduction
merged_seurat[["percent.mt"]] <- PercentageFeatureSet(merged_seurat, pattern = "^mt-")
plot1 <- FeatureScatter(merged_seurat, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(merged_seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

```{r}
merged_seurat <- subset(merged_seurat, subset = nFeature_RNA > 200 & nFeature_RNA < 4400 & percent.mt < 15)
merged_seurat <- Seurat::NormalizeData(merged_seurat, verbose = FALSE) %>% 
  FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>%   
  CellCycleScoring(g2m.features = m.g2m.genes,
                   s.features = m.s.genes,
                   set.ident = FALSE) %>% 
  ScaleData(vars.to.regress = c("S.Score", "G2M.Score")) %>% 
  RunPCA(pc.genes = merged_seurat@var.genes, npcs = 20, verbose = FALSE)

ElbowPlot(merged_seurat)
```

```{r}
# Run harmony integration
merged_seurat_harmony <- merged_seurat %>%
  RunHarmony("genotype")

ElbowPlot(merged_seurat_harmony)

merged_seurat_harmony <- RunUMAP(merged_seurat_harmony, reduction='harmony', 
                                 dims = 1:8)
merged_seurat_harmony <- FindNeighbors(merged_seurat_harmony, reduction='harmony', 
                                       dims = 1:8)
merged_seurat_harmony <- FindClusters(merged_seurat_harmony, resolution = 0.7)
```


```{r}
# Doublet detection and removal
sweep.list <- paramSweep_v3(plaque_harmony, PCs = 1:min.pc, num.cores = detectCores() - 1)
sweep.stats <- summarizeSweep(sweep.list)
bcmvn <- find.pK(sweep.stats)
nExp <- round(0.03*nrow(plaque_harmony@meta.data)) 
plaque_doublets <- doubletFinder_v3(plaque_harmony, PCs = 1:8, pN = 0.25, pK = 0.09, nExp = nExp, reuse.pANN = FALSE, sct = FALSE)
plaque_harmony = plaque_doublets[, plaque_doublets@meta.data[, DF.name] == "Singlet"]
```

